packages/consensus/src/validation/script.cairo

```diff
@@ -0,0 +1,187 @@
+    if header.version >= @3_u32 && block_height >= @BIP_66_BLOCK_HEIGHT {
+        script_flags += ScriptFlags::ScriptVerifyDERSignatures.into();
+    }
+
+    // Enforce CHECKLOCKTIMEVERIFY for block versions 4+
+    // This is part of BIP0065.
+    if header.version >= @4_u32 && block_height >= @BIP_65_BLOCK_HEIGHT {
+        script_flags += ScriptFlags::ScriptVerifyCheckLockTimeVerify.into();
+    }
+
+    // Enforce CHECKSEQUENCEVERIFY if the CSV soft-fork is now active
+    //  This is part of BIP0112.
+    if block_height >= @BIP_112_BLOCK_HEIGHT {
+        script_flags += ScriptFlags::ScriptVerifyCheckSequenceVerify.into();
+    }
+
+    // Enforce the segwit soft-fork
+    // This is part of BIP0141.
+    if block_height >= @BIP_141_BLOCK_HEIGHT {
+        script_flags += ScriptFlags::ScriptVerifyWitness.into();
+        script_flags += ScriptFlags::ScriptStrictMultiSig.into();
+    }
+
+    // Enforce the taproot soft-fork
+    // This is part of BIP0341.
+    if block_height >= @BIP_341_BLOCK_HEIGHT {
+        script_flags += ScriptFlags::ScriptVerifyTaproot.into();
+    }
+
+    script_flags
+}
+
+fn parse_short_string(short: felt252) -> ByteArray {
+    let mut f: u256 = short.into();
+    let mut l = 0;
+    while f != 0 {
+        f = f / 256;
+        l += 1;
+    };
+    let mut parsed: ByteArray = Default::default();
+    parsed.append_word(short, l);
+    parsed
+}
+
+fn validate_authorization(header: @Header, tx: @Transaction, tx_idx: u32) -> Result<(), ByteArray> {
+    let cache = HashCacheImpl::new(tx);
+
+    let mut result: Option<ByteArray> = Option::None;
+    let inputs_len = (*tx.inputs).len();
+    for input_idx in 0
+        ..inputs_len {
+            let previous_output = *tx.inputs[input_idx].previous_output;
+            let mut engine = EngineImpl::new(
+                previous_output.data.pk_script,
+                tx,
+                input_idx,
+                script_flags(header, tx),
+                Into::<u64, i128>::into(previous_output.data.value).try_into().unwrap(),
+                @cache
+            )
+                .unwrap(); //TODO: handle error
+
+            match engine.execute() {
+                Result::Ok(_) => { break; }, // TODO: verify this is correct
+                Result::Err(err) => {
+                    result =
+                        Option::Some(
+                            format!(
+                                "Script validation failed on tx_idx: {}, input_idx: {}: {}",
+                                tx_idx,
+                                input_idx,
+                                parse_short_string(err)
+                            )
+                        );
+                    break;
+                }
+            }
+        };
+
+    match result {
+        Option::Some(err) => Result::Err(err),
+        Option::None => Result::Ok(())
+    }
+}
+
+pub fn validate_authorizations(header: @Header, txs: Span<Transaction>) -> Result<(), ByteArray> {
```

Comment:
Can you pls add docstring.
Also why not `validate_scripts`?
Would it make sense to do validation in the `validate_transaction`?

Reply:
> Would it make sense to do validation in the `validate_transaction`?
Maybe, but the cost of having a separate loop is negligible.